pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+	}
+
+	routes, err := m.rpdb.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+		Table: oldTable,
+	}, netlink.RT_FILTER_TABLE)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list routes associated with rule: %w", err)
+	}
+
+	var stack revert.RevertStack
+
+	revert, err := m.copyRoutes(routes, oldTable, newTable)
+	stack.Extend(revert)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new routes: %w", err)
+	}
+
+	// We don't need the revert stack return value because the next operation
+	// to delete the rule will not revert the stack. See below comment on why.
+	_, err = m.createNewRule(
+		rule,
+		linux_defaults.RulePriorityEgress,
+		newTable,
+	)
+	if err != nil {
+		if revErr := stack.Revert(); revErr != nil {
+			scopedLog.WithError(err).WithField("revertError", revErr).Warn(downgradeRevertWarning)
+		}
+
+		return nil, fmt.Errorf("failed to create new rule: %w", err)
+	}
+
+	if err := m.rpdb.RuleDel(&rule); err != nil {
+		// We avoid reverting and returning an error here because the newer
+		// datapath is already in-place. See long comment describing the
+		// migration in MigrateENIDatapath().
+		scopedLog.WithError(err).Warn(downgradeFailedRuleDeleteWarning)
+		return nil, nil
+	}
+
+	return routes, nil
+}
+
+const (
+	upgradeRevertWarning = "Reverting the new ENI datapath failed. However, the previous datapath is still intact. " +
+		"Endpoint connectivity should not be affected. It is advised to retry the migration."
+	downgradeRevertWarning = "Reverting the new ENI datapath failed. However, both the new and previous datapaths are still intact. " +
+		"Endpoint connectivity should not be affected. It is advised to retry the migration."
+	downgradeFailedRuleDeleteWarning = "Downgrading the datapath has succeeded, but failed to cleanup the original datapath. " +
+		"It is advised to manually remove the old rule."
+)
+
+// retrieveTableIDFromIfIndex computes the correct table ID based on the
+// ifindex provided. The table ID is comprised of the number associated with an
+// ENI device that corresponds to the ifindex, plus the specific table offset
+// value.
+func (m *migrator) retrieveTableIDFromIfIndex(ifindex int) (int, error) {
+	link, err := m.rpdb.LinkByIndex(ifindex)
+	if err != nil {
+		return -1, fmt.Errorf("failed to find link by index: %w", err)
+	}
+
+	mac := link.Attrs().HardwareAddr.String()
+	ifaceNum, err := m.getter.GetInterfaceNumberByMAC(mac)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get interface-number by MAC %q: %w", mac, err)
+	}
+
+	// This is guaranteed to avoid conflicting with the main routing table ID
+	// (253-255) because the maximum number of ENI devices on a node is 15 (see
+	// pkg/aws/eni/limits.go). Because the interface number is monotonically
+	// increasing and the lowest available number is reused when devices are
+	// added / removed. This means that the max possible table ID is 25.
+	return linux_defaults.RouteTableInterfacesOffset + ifaceNum, nil
```

Comment:
Is it worth adding a check here that the returned table ID does not actually conflict with the main routing table? Of course, right now the maximum number of ENI devices on a node is 15, but this may change in the future. Adding a check here would catch the problem early rather than inadvertently colliding with the main routing table when a machine in the future gets its 243rd ENI.

Reply:
Added a check which will warn the user in this case.